doc/source/admin/container_resolvers.rst

```diff
@@ -0,0 +1,276 @@
+to obtain a container description by sequentially executing the container
+resolvers that are configured for the Galaxy instance (see below).
+The job is then executed using the description returned by the first successful
+container resolver.
+If all configured container resolvers failed, i.e. no container description
+could be obtained, the tool is by default excuted using 
+:doc:`standard dependency resolvers <dependency_resolvers>`, e.g. ``conda``.
+Alternatively, if the excution environment specifies
+`require_container <https://github.com/galaxyproject/galaxy/blob/0742d6e27702c60d1b8fe358ae03a267e3f252c3/lib/galaxy/config/sample/job_conf.sample.yml#L528>`_
+the job fails in this case (TODO to be tested? or is it already).
+
+Configuration:
+--------------
+
+The list of container resolvers is defined using YAML that can be
+specified in an extra file (``container_resolvers_config_file``) or
+inline the Galaxy configuration (``container_resolvers``). 
+During the container resolution the resolvers specified in this
+list are sequentially applied stopping at the first resolver that
+yields a container description. 
+
+The YAML defines a list of container resolvers and their properties.
+This might look as follows:
+
+.. code-block:: yaml
+
+   - type: "cached_explicit_singularity"
+   - type: "cached_mulled_singularity"
+   - type: "mulled_singularity"
+     auto_install: False
+
+would define a list of three container resolvers, and the last setting an option
+of this resolver.
+
+The available container resolver types are described below.
+The properties depend on the resolver type (see documentation
+of the resolvers). Properties that do not apply to the resolver
+are ignored.
+
+.. note::
+
+   Its's important to note that similarities in the names not necessarily
+   imply any similarity in the function of the container resolvers.
+
+Main resolver types:
+--------------------
+
+The main types of container resolvers follow this naming scheme: 
+``[cached_][explicit,mulled][_singularity]``. That is
+
+- a container resolver is either ``explicit`` or ``mulled``
+- cached if it is prefixed with ``cached_`` and non-cached otherwise. 
+- yield a container discription suitable for singularity if
+  suffixed by ``_singularity`` and docker otherwise.
+
+1. docker vs singularity
+
+Galaxy can execute tools in containers using  ``docker`` or ``singularity``.
+The corresponding container resolvers yield container descriptions suitable
+for the corresponding "executor".
+Note that for the execution with ``singularity`` Galaxy relies mostly on
+docker containers that are either executed directly or are converted
+to singularity images (except for explicit container requirements of
+``type="singularity"``).
+
+There are important differences between Galaxy's cached docker and singularity
+container resolvers. The caching mechanism essentially executes a
+``docker pull`` or ``singularity pull``, respectively. For docker this creates
```

Comment:
Is the `singularity pull` currently being called? Last time I setup something relying on singularity, I had to have a mechanism of getting the tool containers apriori I think. Or maybe I did this to avoid issues with concurrent pulls when dataset collections where the first ones to request the pull of an image... not sure. Might be good to know as well if concurrent `singularity pull` calls are somehow avoided currently.

Reply:
Some of them do:

- `cached_explicit_singularity` triggers a pull over [here](https://github.com/galaxyproject/galaxy/blob/c517e805771cc16807dfe675075a13fe6343f01f/lib/galaxy/tool_util/deps/container_resolvers/explicit.py#L106)
- `mulled` and `mulled_singularity` trigger pulls [here](https://github.com/galaxyproject/galaxy/blob/c517e805771cc16807dfe675075a13fe6343f01f/lib/galaxy/tool_util/deps/container_resolvers/mulled.py#L592) which are then executed in the respective methods [here for `mulled`](https://github.com/galaxyproject/galaxy/blob/c517e805771cc16807dfe675075a13fe6343f01f/lib/galaxy/tool_util/deps/container_resolvers/mulled.py#L540) and [here for `mulled_singularity`](https://github.com/galaxyproject/galaxy/blob/c517e805771cc16807dfe675075a13fe6343f01f/lib/galaxy/tool_util/deps/container_resolvers/mulled.py#L632)

The pull happens when `install=True` (which is apparently the case when called during job preparation and when build is triggered from the container dependencies tab in the admin UI) ... seems that the value of `auto_intstall` is important here as well since it determines if the container description uses the URI / the path to the image (but I don't fully understand this yet).

The `cached_mulled*` only lookup the cache.